294 fallbackroute should return opt

ekondrashev opened the pull request:
Please review this PR for #294 

It switches `Fallback.route` to use `Opt` instead of `Iterator`

davvd commented:
@ekondrashev Thanks for your pull request, let me find someone who can review it

davvd commented:
@pinaf this pull request is for you, please review

ekondrashev commented:
@pinaf seems like the PR was lost, could you please provide your comments?

pinaf commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -96,20 +93,20 @@ public boolean fits(final Integer status) {
-                        }
+                        rsp = fallback.route(req);
+                    } else {
+                        rsp = new Opt.Empty<Response>();
```
@ekondrashev move this assignment to line 103 above.

pinaf commented on src/test/java/org/takes/facets/fallback/TkFallbackTest.java:
```diff
@@ -60,10 +59,10 @@ public void fallsBack() throws IOException {
-                            ).iterator();
+                        public Opt<Response> route(final RqFallback req) {
+                            return new Opt.Single<Response>(
+                                    new RsText(req.throwable().getMessage())
```
@ekondrashev too much indentation here

pinaf commented:
@ekondrashev 2 minor comments above

pinaf commented:
@ekondrashev we need a merge with master

ekondrashev commented:
@pinaf I've merged with master and fixed all the comments.
I've also added ClassDataAbstractionCouplingCheck check style warning suppression for FbStatus.java not sure if I need to create a ticket to resolve it as puzzle.

Please let me know if PR needs another changes.

pinaf commented on .rultor.yml:
```diff
@@ -34,7 +34,9 @@ merge:
     mvn clean
     pdd --source=$(pwd) --verbose --file=/dev/null
   commanders:
+  - caarlos0
```
@ekondrashev this doesn't look good.. looks like your merge was not done right

pinaf commented on README.md:
````diff
@@ -419,7 +420,7 @@ new TkFork(
 
 ```java
-Href href = new RqHref(request).href();
+Href href = new RqHref.Base(request).href();
````
@ekondrashev this doesn't look good.. looks like your merge was not done right

pinaf commented:
@ekondrashev looks like your merge was not done correctly. there are multiple files that should not have any changes but do..

ekondrashev commented:
@pinaf sorry for that, fixed.

pinaf commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -43,8 +40,10 @@
  * <p>The class is immutable and thread-safe.
  *
  * @author Yegor Bugayenko (yegor@teamed.io)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
@ekondrashev this is duplicated

pinaf commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -43,8 +40,10 @@
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  * @version $Id$
  * @since 0.13
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
@ekondrashev are you sure we even need this suppression? it looks to me that you actually reduced the coupling (the imports above were reduced)

pinaf commented:
@ekondrashev thanks. 2 comments

ekondrashev commented:
@pinaf removed the suppression. As for the second comment, yes it is needed, qulice gives 
```
[ERROR] /src/main/java/org/takes/facets/fallback/FbStatus.java[46]: Class Data Abstraction Coupling is 8 (max allowed is 7) classes [Condition, Fallback, Opt.Empty, Opt.Single, RsWithBody, RsWithStatus, RsWithType, TkFixed]. (ClassDataAbstractionCouplingCheck)
``` 
without it.

pinaf commented:
@ekondrashev thanks

pinaf commented:
@rultor merge

rultor commented:
> @rultor merge

@pinaf Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/main/java/org/takes/facets/fallback/FbChain.java:
```diff
@@ -55,21 +53,22 @@ public FbChain(final Fallback... fallbacks) {
-                            rsp.add(iter.next());
+                        final Opt<Response> opt = fbk.route(req);
+                        if (opt.has()) {
+                            rsp = new Opt.Single<Response>(opt.get());
```
@ekondrashev why not just `rsp = opt`?

yegor256 commented:
@ekondrashev one comment above

ekondrashev commented:
@yegor256 right, fixed

ekondrashev commented:
@yegor256 could you please take a look at this PR?

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/3957-107650129)

rultor commented:
> @rultor try to merge

@yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/3957-107650129) (spent 9min)

```

Results :

Failed tests: 
  BkParallelTest.requestsAreParallel:123 
Expected: <0L>
     but: was <1L>

Tests run: 222, Failures: 1, Errors: 0, Skipped: 5

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:50 min
[INFO] Finished at: 2015-06-01T17:46:43+00:00
[INFO] Final Memory: 24M/371M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project takes: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/r/repo/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project takes: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution(SurefireHelper.java:82)
	at org.apache.maven.plugin.surefire.SurefirePlugin.handleSummary(SurefirePlugin.java:195)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:861)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:729)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 19 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

ekondrashev commented:
@yegor256 could you please give an advice on the error above?

yegor256 commented:
@ekondrashev we tried to fix that in #309, but failed, as I see...

yegor256 commented:
@rultor try to merge again

rultor commented:
> @rultor try to merge again

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/3957-109363038)

rultor commented:
> @rultor try to merge again

@yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/3957-109363038) (spent 10min)

```

Results :

Failed tests: 
  BkParallelTest.requestsAreParallel:127 
Expected: <0L>
     but: was <1L>

Tests run: 226, Failures: 1, Errors: 0, Skipped: 4

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:58 min
[INFO] Finished at: 2015-06-05T17:09:10+00:00
[INFO] Final Memory: 25M/443M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project takes: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/r/repo/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project takes: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution(SurefireHelper.java:82)
	at org.apache.maven.plugin.surefire.SurefirePlugin.handleSummary(SurefirePlugin.java:195)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:861)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:729)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 19 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

ekondrashev commented:
@yegor256 can we perhaps disable the test and reopen the #309?

yegor256 commented:
@ekondrashev #309 is already closed and paid, no way to reopen it. But you can report a new ticket, explain the problem there and wait here until it's fixed

ekondrashev commented:
@yegor256 done, plz see #333 
@davvd this PR merge is blocked by #333 hence original defect is also blocked, will comment on it also

pinaf commented:
@davvd we're waiting for #333 here

pinaf commented:
@yegor256 since the test is unstable, we could try to merge again I supposed. Maybe we get lucky this time :)

yegor256 commented:
@rultor try to merge again

rultor commented:
> @rultor try to merge again

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/3957-109435445)

rultor commented:
> @rultor try to merge again

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/3957-109435445) (took me 11min)

davvd commented:
> @davvd we're waiting for #333 here

@pinaf right, let's wait for #333

ekondrashev commented:
@davvd I believe there is no reason to wait for #333 now, since the PR was merged and initial ticket is closed already.

davvd commented:
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

davvd commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/3957-109732248)

rultor commented:
> @rultor please deploy

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/3957-109732248) (took me 9min)

elenavolokhova commented:
@davvd Quality is good here.

davvd commented:
> @davvd Quality is good here.

@elenavolokhova thanks for the review and a good remark

davvd commented:
@pinaf since quality is good, 10 mins paid to @yegor256 (our architect), transaction `000-6a463b2b`

davvd commented:
@pinaf 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 58704981; Thanks a lot, I just topped your account for **23 mins**, transaction ID `AP-29N57172GJ6916946`, total time was 426 hours and 41 mins; extra minutes for review comments (c=8); +23 added to your rating, current score is: [+7665](http://www.netbout.com/b/35128?open=rating)
